python_modules/libraries/dagster-components/dagster_components/lib/definitions_component.py

```diff
@@ -0,0 +1,49 @@
+from pathlib import Path
+from typing import Optional
+
+from dagster._core.definitions.definitions_class import Definitions
+from dagster._core.definitions.module_loaders.load_defs_from_module import (
+    load_definitions_from_module,
+)
+from dagster._seven import import_module_from_path
+from dagster._utils import pushd
+from pydantic import BaseModel
+from typing_extensions import Self
+
+from dagster_components import Component, ComponentGenerateRequest, ComponentLoadContext, component
+
+
+class DefinitionsGenerateParams(BaseModel):
+    definitions_path: Optional[str] = None
+
+
+class DefinitionsParamSchema(BaseModel):
+    definitions_path: Optional[str] = None
+
+
+@component(name="definitions")
+class DefinitionsComponent(Component):
+    def __init__(self, definitions_path: Path):
+        self.definitions_path = definitions_path
+
+    generate_params_schema = DefinitionsGenerateParams
+    params_schema = DefinitionsParamSchema
+
+    @classmethod
+    def load(cls, context: ComponentLoadContext) -> Self:
+        # all paths should be resolved relative to the directory we're in
+        loaded_params = context.load_params(cls.params_schema)
+
+        return cls(definitions_path=Path(loaded_params.definitions_path or "definitions.py"))
+
+    def build_defs(self, context: ComponentLoadContext) -> Definitions:
+        with pushd(str(context.path)):
+            module = import_module_from_path("definitions", str(self.definitions_path))
```

Comment:
@smackesey one thing I want to call out here is what our convention for module names should be.

Reply:
How do you feel about making `components` itself a Python package? This is a little weird because it contains all kinds of non-python stuff but it means that Python modules in component instances have a home in the hierarchy, and we are already putting `components` inside a package anyway.

So the module name here would be `<code-loc-name>.components.<component-name>.definitions`

Reply:
Having now seen that previously merged PR where the components folder is parameterizable I can see that that won't work.

I haven't thought through whether we need module caching, but I think the best thing to do for now is to:

- Continue using a basic name like `definitions` as here
- Ensure the method used to load the module skips the cache
- Ensure this `build_defs` function has an explanatory comment explaining how the basic name works because we're not using the cache

Alternatively we could assign a name like `__dagster_components__.<component-name>.definitions`, which might make debugging easier.